### ISSUE TITLE
`INIT_ARRAY` is forced into the binary + Strong coupling between `syscalls` and `impl::process` 

### ISSUE THREAD

**Issue by AsafFisher** [bytecodealliance/rustix#338]
Some features of `imp::process` such as `auxv.rs`'s `INIT_ARRAY` are forced into the binary. I think their inclusion should be a feature.
It will further separate the dependencies between raw syscalls and usermode tricks.
Such a feature would be called `startfile` or `disable-initfini-array` right?

I did not dive too deep into this. Let me know what you think...

P.S. At first I tried to just comment out `::process` but turns out `syscalls.rs` uses it. 

https://github.com/bytecodealliance/rustix/blob/67fb48927fd0aedbd0d5976ad14871aa873647e1/src/imp/linux_raw/fs/syscalls.rs#L1318

**Comment by AsafFisher** [comment]
Basically I got this in my `no-std`, `pic` project:
```rust
note: rust-lld: error: undefined symbol: __environ
          >>> referenced by shellcode.c5f0cd7e-cgu.0
          >>>               /home/a/workspace/dbg/target/aarch64-unknown-none/release/deps/shellcode-32bd3c403fa05c97.shellcode.c5f0cd7e-cgu.0.rcgu.o:(rustix::imp::process::auxv::INIT_ARRAY::function::h6bb1669a717e91aa)
```

**Comment by sunfishcode** [comment]
https://github.com/bytecodealliance/rustix/pull/339 is a start. That would mean that instead of depending on `__environ`, rustix would need users using `target_os = "none"` to call `rustix::process::init` some time early in process startup. For example, [this is what mustang does](https://github.com/sunfishcode/mustang/blob/main/origin/src/program.rs#L74) in its program startup code. Is that something you could do?

Beyond that, I'm also considering adding something like an `"conf"` (eymology: "sysconf" without the "sys" because it's about process configuration rather than "system" configuration) cargo feature, and guarding `page_size()` and others with it, so that users not using those function wouldn't need the initialization at all.

https://github.com/bytecodealliance/rustix/pull/339  is an initial PR implementing these.

**Comment by AsafFisher** [comment]
Hmmm it doesnt really make sense for users that use rustix to set `target_os = "none"` isnt it?

**Comment by sunfishcode** [comment]
Isn't that what you yourself are doing? :smile: The error message says aarch64-unknown-none.

**Comment by AsafFisher** [comment]
Well I am actually doing a trick where I built a custom `.json` target file, there I set the `target_os` to Linux

**Comment by sunfishcode** [comment]
I'm not clear on what's happening here. The `__environ` symbol is [only used in non-"gnu" configurations](https://github.com/bytecodealliance/rustix/blob/main/src/imp/linux_raw/process/auxv.rs#L124), however the error message above mentioned an undefined reference to `__environ`. Is the aarch64 config using a different "os" or "env"?

**Comment by sunfishcode** [comment]
Ok, that makes sense. Then I'd expect #339 will fix the issue, because if you declare your target_env to be "gnu", rustix will assume it can use gnu's `.init_array` convention and avoid using `__environ`, and if you declare your target_env to be "none", rustix will export an `init` function and `debug_assert` if you don't call it.

And on top of that, splitting off "conf" as a separate module should mean that you don't need any of this if you don't need "conf" or `clock_gettime` or 32-bit x86 support.

**Comment by sunfishcode** [comment]
#339 is now released in 0.35. Can you confirm whether this fixes the issue for you?

**Comment by sunfishcode** [comment]
I expect this is fixed in #339. Please reopen or file a new issue if there's anything that doesn't work.
